diff --git a/apps/dgst.c b/apps/dgst.c
index e595f7d..bd1d903 100644
--- a/apps/dgst.c
+++ b/apps/dgst.c
@@ -566,7 +566,7 @@ int do_fp(BIO *out, unsigned char *buf, BIO *bp, int sep, int binout,
         for (i = 0; i < (int)len; i++)
             BIO_printf(out, "%02x", buf[i]);
 
-        BIO_printf(out, " *%s\n", file);
+        BIO_printf(out, "  %s\n", file);
         OPENSSL_free((char *)file);
     } else {
         if (sig_name != NULL) {
