diff --git a/library/unwind/build.rs b/library/unwind/build.rs
index ab09a6e3..69c9320d 100644
--- a/library/unwind/build.rs
+++ b/library/unwind/build.rs
@@ -14,7 +14,7 @@ fn main() {
     } else if target.contains("linux") {
         if target.contains("musl") {
             // linking for musl is handled in lib.rs
-            llvm_libunwind::compile();
+            println!("cargo:rustc-link-lib=unwind");
         } else if !target.contains("android") {
             println!("cargo:rustc-link-lib=gcc_s");
         }
