From 63483d7e22efe320b40e64021c6b19efaa775002 Mon Sep 17 00:00:00 2001
From: Michael Forney <mforney@mforney.org>
Date: Fri, 7 Jun 2019 21:51:51 -0700
Subject: [PATCH] objtool: Use Elf_Scn typedef instead of assuming tag name

---
 tools/objtool/elf.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/objtool/elf.c b/tools/objtool/elf.c
index 8d78c8a0660072..96f1c34d9120b2 100644
--- a/tools/objtool/elf.c
+++ b/tools/objtool/elf.c
@@ -468,7 +468,7 @@ struct section *elf_create_section(struct elf *elf, const char *name,
 {
 	struct section *sec, *shstrtab;
 	size_t size = entsize * nr;
-	struct Elf_Scn *s;
+	Elf_Scn *s;
 	Elf_Data *data;
 
 	sec = malloc(sizeof(*sec));
