From 52f1c8b97b8adce9d579767fbefeeebc27389de8 Mon Sep 17 00:00:00 2001
From: Drew DeVault <sir@cmpwn.com>
Date: Fri, 7 Dec 2018 21:25:08 -0500
Subject: [PATCH] Add riscv64 to apk_defines.h

---
 src/apk_defines.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/apk_defines.h b/src/apk_defines.h
index c7ab7e6..8bbd69f 100644
--- a/src/apk_defines.h
+++ b/src/apk_defines.h
@@ -113,6 +113,10 @@ extern char **apk_argv;
 #define APK_DEFAULT_ARCH	"mips"
 #elif defined(__mips__) && __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
 #define APK_DEFAULT_ARCH	"mipsel"
+#elif defined(__riscv) && __riscv_xlen == 32
+#define APK_DEFAULT_ARCH	"riscv32"
+#elif defined(__riscv) && __riscv_xlen == 64
+#define APK_DEFAULT_ARCH	"riscv64"
 #else
 #error APK_DEFAULT_ARCH not detected for this architecture
 #endif
-- 
2.19.2

