diff --git a/libfetch/common.c b/libfetch/common.c
index e237828..ee48af9 100644
--- a/libfetch/common.c
+++ b/libfetch/common.c
@@ -184,6 +184,8 @@ fetch_default_port(const char *scheme)
 		return (FTP_DEFAULT_PORT);
 	if (strcasecmp(scheme, SCHEME_HTTP) == 0)
 		return (HTTP_DEFAULT_PORT);
+	if (strcasecmp(scheme, SCHEME_HTTPS) == 0)
+		return (HTTPS_DEFAULT_PORT);
 	return (0);
 }
 
diff --git a/libfetch/common.h b/libfetch/common.h
index d7ad1d6..fc78f16 100644
--- a/libfetch/common.h
+++ b/libfetch/common.h
@@ -34,6 +34,7 @@
 
 #define FTP_DEFAULT_PORT	21
 #define HTTP_DEFAULT_PORT	80
+#define HTTPS_DEFAULT_PORT	443
 #define FTP_DEFAULT_PROXY_PORT	21
 #define HTTP_DEFAULT_PROXY_PORT	3128
 
