diff --git a/src/apk_defines.h b/src/apk_defines.h
index a525d33..09de79f 100644
--- a/src/apk_defines.h
+++ b/src/apk_defines.h
@@ -93,13 +93,13 @@ extern char **apk_argv;
 #define APK_DEFAULT_ARCH	"x86"
 #define APK_DEFAULT_ENDIAN	"little"
 #elif defined(__powerpc__) && !defined(__powerpc64__)
-#define APK_DEFAULT_ARCH	"ppc"
+#define APK_DEFAULT_ARCH	"powerpc"
 #define APK_DEFAULT_ENDIAN	"big"
 #elif defined(__powerpc64__) && __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
-#define APK_DEFAULT_ARCH	"ppc64"
+#define APK_DEFAULT_ARCH	"powerpc64"
 #define APK_DEFAULT_ENDIAN	"big"
 #elif defined(__powerpc64__) && __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
-#define APK_DEFAULT_ARCH	"ppc64le"
+#define APK_DEFAULT_ARCH	"powerpc64le"
 #define APK_DEFAULT_ENDIAN	"little"
 #elif defined(__arm__) && defined(__ARM_PCS_VFP) && __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__ && __ARM_ARCH>=7
 #define APK_DEFAULT_ARCH	"armv7"
