diff --git a/engine/compiler/compiler.go b/engine/compiler/compiler.go
index 81782d8..5025169 100644
--- a/engine/compiler/compiler.go
+++ b/engine/compiler/compiler.go
@@ -35,7 +35,7 @@ var random = func() string {
 // with privileged capabilities in order to run Docker
 // in Docker.
 var Privileged = []string{
-	"plugins/docker",
+	"abyssos/junk:plugins-docker",
 	"plugins/acr",
 	"plugins/ecr",
 	"plugins/gcr",
