diff --git a/include/private/gcconfig.h b/include/private/gcconfig.h
index 348255d..b578783 100644
--- a/include/private/gcconfig.h
+++ b/include/private/gcconfig.h
@@ -2289,7 +2289,7 @@ EXTERN_C_BEGIN
 #     if defined(HOST_ANDROID)
 #       define SEARCH_FOR_DATA_START
 #     else
-        extern int __data_start[];
+        extern int __data_start[] __attribute__((__weak__));
 #       define DATASTART ((ptr_t)__data_start)
 #     endif
       extern int _end[];
@@ -2879,7 +2879,7 @@ EXTERN_C_BEGIN
 #   define ALIGNMENT (CPP_WORDSZ/8)
 #   ifdef LINUX
 #     define OS_TYPE "LINUX"
-      extern int __data_start[];
+      extern int __data_start[] __attribute__((__weak__));
 #     define DATASTART ((ptr_t)__data_start)
 #     define LINUX_STACKBOTTOM
 #     define DYNAMIC_LOADING
