diff --git a/components/cli/scripts/build/.variables b/components/cli/scripts/build/.variables
index b8ebe66..89bd7a9 100755
--- a/components/cli/scripts/build/.variables
+++ b/components/cli/scripts/build/.variables
@@ -4,7 +4,7 @@ set -eu
 PLATFORM=${PLATFORM:-}
 VERSION=${VERSION:-"unknown-version"}
 GITCOMMIT=${GITCOMMIT:-$(git rev-parse --short HEAD 2> /dev/null || true)}
-BUILDTIME=${BUILDTIME:-$(date --utc --rfc-3339 ns 2> /dev/null | sed -e 's/ /T/')}
+BUILDTIME=${BUILDTIME:-$(date --utc 2>/dev/null)}
 
 PLATFORM_LDFLAGS=
 if test -n "${PLATFORM}"; then
