diff --git a/components/engine/container/container.go b/components/engine/container/container.go
index 71a8a1b..8cf2800 100644
--- a/components/engine/container/container.go
+++ b/components/engine/container/container.go
@@ -21,7 +21,6 @@ import (
 	"github.com/docker/docker/container/stream"
 	"github.com/docker/docker/daemon/exec"
 	"github.com/docker/docker/daemon/logger"
-	"github.com/docker/docker/daemon/logger/jsonfilelog"
 	"github.com/docker/docker/daemon/logger/local"
 	"github.com/docker/docker/daemon/network"
 	"github.com/docker/docker/errdefs"
@@ -376,16 +375,7 @@ func (container *Container) StartLogger() (logger.Logger, error) {
 		DaemonName:          "docker",
 	}
 
-	// Set logging file for "json-logger"
-	// TODO(@cpuguy83): Setup here based on log driver is a little weird.
 	switch cfg.Type {
-	case jsonfilelog.Name:
-		info.LogPath, err = container.GetRootResourcePath(fmt.Sprintf("%s-json.log", container.ID))
-		if err != nil {
-			return nil, err
-		}
-
-		container.LogPath = info.LogPath
 	case local.Name:
 		// Do not set container.LogPath for the local driver
 		// This would expose the value to the API, which should not be done as it means
