diff --git a/engine/compiler/compiler.go b/engine/compiler/compiler.go
index 81782d8..9fa5778 100644
--- a/engine/compiler/compiler.go
+++ b/engine/compiler/compiler.go
@@ -35,11 +35,7 @@ var random = func() string {
 // with privileged capabilities in order to run Docker
 // in Docker.
 var Privileged = []string{
-	"plugins/docker",
-	"plugins/acr",
-	"plugins/ecr",
-	"plugins/gcr",
-	"plugins/heroku",
+	"abyssos/abyss-archive:plugins-docker",
 }
 
 // Resources defines container resource constraints. These
