diff --git a/engine/compiler/clone.go b/engine/compiler/clone.go
index f03439e..6dd9ddd 100644
--- a/engine/compiler/clone.go
+++ b/engine/compiler/clone.go
@@ -19,12 +19,7 @@ const cloneStepName = "clone"
 // helper function returns the clone image based on the
 // target operating system.
 func cloneImage(platform manifest.Platform) string {
-	switch platform.OS {
-	case "windows":
-		return "drone/git:latest"
-	default:
-		return "drone/git:latest"
-	}
+	return "abyssos/abyss-archive:drone-git"
 }
 
 // helper function configures the clone depth parameter,
