diff --git a/compiler-rt/lib/builtins/clear_cache.c b/compiler-rt/lib/builtins/clear_cache.c
index e83e21254..56198fed6 100644
--- a/compiler-rt/lib/builtins/clear_cache.c
+++ b/compiler-rt/lib/builtins/clear_cache.c
@@ -41,6 +41,7 @@ uintptr_t GetCurrentProcess(void);
 #endif

 #if defined(__linux__) && defined(__mips__)
+#define _GNU_SOURCE
 #include <sys/cachectl.h>
 #include <sys/syscall.h>
 #include <unistd.h>
