diff --git a/lib/Driver/ToolChains/Arch/Mips.cpp b/lib/Driver/ToolChains/Arch/Mips.cpp
index b512ff6..172937a 100644
--- a/lib/Driver/ToolChains/Arch/Mips.cpp
+++ b/lib/Driver/ToolChains/Arch/Mips.cpp
@@ -172,10 +172,10 @@ mips::FloatABI mips::getMipsFloatABI(const Driver &D, const ArgList &Args) {
 
   // If unspecified, choose the default based on the platform.
   if (ABI == mips::FloatABI::Invalid) {
-    // Assume "hard", because it's a default value used by gcc.
+    // Assume "hard", because it's a default value used by gcc. (no it isn't)
     // When we start to recognize specific target MIPS processors,
     // we will be able to select the default more correctly.
-    ABI = mips::FloatABI::Hard;
+    ABI = mips::FloatABI::Soft;
   }
 
   assert(ABI != mips::FloatABI::Invalid && "must select an ABI");
