diff --git a/clang/lib/Driver/ToolChains/Arch/Mips.cpp b/clang/lib/Driver/ToolChains/Arch/Mips.cpp
index 7b4dd703c..5011acd81 100644
--- a/clang/lib/Driver/ToolChains/Arch/Mips.cpp
+++ b/clang/lib/Driver/ToolChains/Arch/Mips.cpp
@@ -166,7 +166,7 @@ mips::FloatABI mips::getMipsFloatABI(const Driver &D, const ArgList &Args,
                 .Default(mips::FloatABI::Invalid);
       if (ABI == mips::FloatABI::Invalid && !StringRef(A->getValue()).empty()) {
         D.Diag(clang::diag::err_drv_invalid_mfloat_abi) << A->getAsString(Args);
-        ABI = mips::FloatABI::Hard;
+        ABI = mips::FloatABI::Soft;
       }
     }
   }
@@ -180,7 +180,7 @@ mips::FloatABI mips::getMipsFloatABI(const Driver &D, const ArgList &Args,
       // Assume "hard", because it's a default value used by gcc.
       // When we start to recognize specific target MIPS processors,
       // we will be able to select the default more correctly.
-      ABI = mips::FloatABI::Hard;
+      ABI = mips::FloatABI::Soft;
     }
   }
 
