diff --git a/mk/os.Linux.mk b/mk/os.Linux.mk
index a8208c8..1ab0ca1 100644
--- a/mk/os.Linux.mk
+++ b/mk/os.Linux.mk
@@ -5,17 +5,12 @@
 OS_DISTRIBUTION         != lsb_release -s -i || echo unknown
 OS_DISTRIBUTION_VERSION != lsb_release -s -r || echo unknown
 
-.if ${OS_DISTRIBUTION} == "unknown" || \
-    ${OS_DISTRIBUTION_VERSION} == "unknown"
-.error ERROR: Unknown host OS distribution.
-.endif
-
-MKDOC?=		yes	# Build documentation.
+MKDOC?=		no	# Build documentation.
 MKLINT?=	no
 MKPIC?=		no
 MKNOWEB?=	yes	# Build literate programs.
-MKTESTS?=	yes	# Enable the test suites.
-MKTEX?=		yes	# Build TeX-based documentation.
+MKTESTS?=	no	# Enable the test suites.
+MKTEX?=		no	# Build TeX-based documentation.
 
 OBJECT_FORMAT=	ELF	# work around a bug in the pmake package
 
