diff --git a/common/native-elf-format b/common/native-elf-format
index 2bdd914..57f99e2 100755
--- a/common/native-elf-format
+++ b/common/native-elf-format
@@ -37,6 +37,10 @@ $1 ~ "Machine:" {
             elfarch = "EM_MIPS";
         } else if (match($0, ".*[xX]86-64")) {
             elfarch = "EM_X86_64";
+        } else if (match($0, ".*[Aa]rch64")) {
+            elfarch = "EM_X86_64";
+        } else if (match($0, ".*PowerPC")) {
+            elfarch = "EM_PPC64";
         } else {
             elfarch = "unknown";
         }
