diff --git a/shared/osarch/architectures.go b/shared/osarch/architectures.go
index 84d9f6856..5b24e0f43 100644
--- a/shared/osarch/architectures.go
+++ b/shared/osarch/architectures.go
@@ -14,6 +14,10 @@ const (
 	ARCH_64BIT_POWERPC_BIG_ENDIAN    = 6
 	ARCH_64BIT_POWERPC_LITTLE_ENDIAN = 7
 	ARCH_64BIT_S390_BIG_ENDIAN       = 8
+	ARCH_32BIT_MIPS_BIG_ENDIAN       = 9
+	ARCH_32BIT_MIPS_LITTLE_ENDIAN    = 10
+	ARCH_64BIT_MIPS64_BIG_ENDIAN     = 11
+	ARCH_64BIT_MIPS64_LITTLE_ENDIAN  = 12
 )
 
 var architectureNames = map[int]string{
@@ -25,6 +29,10 @@ var architectureNames = map[int]string{
 	ARCH_64BIT_POWERPC_BIG_ENDIAN:    "ppc64",
 	ARCH_64BIT_POWERPC_LITTLE_ENDIAN: "ppc64le",
 	ARCH_64BIT_S390_BIG_ENDIAN:       "s390x",
+	ARCH_32BIT_MIPS_BIG_ENDIAN:       "mips",
+	ARCH_32BIT_MIPS_LITTLE_ENDIAN:    "mipsle",
+	ARCH_64BIT_MIPS64_BIG_ENDIAN:     "mips64",
+	ARCH_64BIT_MIPS64_LITTLE_ENDIAN:  "mips64le",
 }
 
 var architectureAliases = map[int][]string{
@@ -35,6 +43,10 @@ var architectureAliases = map[int][]string{
 	ARCH_32BIT_POWERPC_BIG_ENDIAN:    {"powerpc"},
 	ARCH_64BIT_POWERPC_BIG_ENDIAN:    {"powerpc64", "ppc64"},
 	ARCH_64BIT_POWERPC_LITTLE_ENDIAN: {"ppc64el"},
+	ARCH_32BIT_MIPS_BIG_ENDIAN:       {"mips"},
+	ARCH_32BIT_MIPS_LITTLE_ENDIAN:    {"mipsle"},
+	ARCH_64BIT_MIPS64_BIG_ENDIAN:     {"mips64"},
+	ARCH_64BIT_MIPS64_LITTLE_ENDIAN:  {"mips64le"},
 }
 
 var architecturePersonalities = map[int]string{
@@ -46,6 +58,10 @@ var architecturePersonalities = map[int]string{
 	ARCH_64BIT_POWERPC_BIG_ENDIAN:    "linux64",
 	ARCH_64BIT_POWERPC_LITTLE_ENDIAN: "linux64",
 	ARCH_64BIT_S390_BIG_ENDIAN:       "linux64",
+	ARCH_32BIT_MIPS_BIG_ENDIAN:       "linux32",
+	ARCH_32BIT_MIPS_LITTLE_ENDIAN:    "linux32",
+	ARCH_64BIT_MIPS64_BIG_ENDIAN:     "linux64",
+	ARCH_64BIT_MIPS64_LITTLE_ENDIAN:  "linux64",
 }
 
 var architectureSupportedPersonalities = map[int][]int{
@@ -57,6 +73,10 @@ var architectureSupportedPersonalities = map[int][]int{
 	ARCH_64BIT_POWERPC_BIG_ENDIAN:    {ARCH_32BIT_POWERPC_BIG_ENDIAN},
 	ARCH_64BIT_POWERPC_LITTLE_ENDIAN: {},
 	ARCH_64BIT_S390_BIG_ENDIAN:       {},
+	ARCH_32BIT_MIPS_BIG_ENDIAN:       {},
+	ARCH_32BIT_MIPS_LITTLE_ENDIAN:    {},
+	ARCH_64BIT_MIPS64_BIG_ENDIAN:     {ARCH_32BIT_MIPS_BIG_ENDIAN},
+	ARCH_64BIT_MIPS64_LITTLE_ENDIAN:  {ARCH_32BIT_MIPS_LITTLE_ENDIAN},
 }
 
 const ArchitectureDefault = "x86_64"
