From 041174437c3cfbee3855da10633f142ae0f0b2ae Mon Sep 17 00:00:00 2001
From: Joe Holden <jwh@zorins.us>
Date: Tue, 19 May 2020 18:11:38 +0000
Subject: [PATCH] shared/util_linux.go: cast Rdev uint64 for mips

Signed-off-by: Joe Holden <jwh@zorins.us>
---
 shared/util_linux.go | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/shared/util_linux.go b/shared/util_linux.go
index 47d35a0e30..b37792f7e9 100644
--- a/_dist/src/github.com/lxc/lxd/shared/util_linux.go
+++ b/_dist/src/github.com/lxc/lxd/shared/util_linux.go
@@ -30,8 +30,8 @@ func GetFileStat(p string) (uid int, gid int, major uint32, minor uint32, inode
 	inode = uint64(stat.Ino)
 	nlink = int(stat.Nlink)
 	if stat.Mode&unix.S_IFBLK != 0 || stat.Mode&unix.S_IFCHR != 0 {
-		major = unix.Major(stat.Rdev)
-		minor = unix.Minor(stat.Rdev)
+		major = unix.Major(uint64(stat.Rdev))
+		minor = unix.Minor(uint64(stat.Rdev))
 	}
 
 	return
