diff --git a/lxd/seccomp/cgo.go b/lxd/seccomp/cgo.go
index 96af14b..7607e28 100644
--- a/lxd/seccomp/cgo.go
+++ b/lxd/seccomp/cgo.go
@@ -2,7 +2,7 @@
 
 package seccomp
 
-// #cgo CFLAGS: -std=gnu11 -Wvla -Werror -fvisibility=hidden -Winit-self
+// #cgo CFLAGS: -std=gnu11 -Wvla -fvisibility=hidden -Winit-self
 // #cgo CFLAGS: -Wformat=2 -Wshadow -Wendif-labels -fasynchronous-unwind-tables
 // #cgo CFLAGS: -pipe --param=ssp-buffer-size=4 -g -Wunused
 // #cgo CFLAGS: -Werror=implicit-function-declaration
