diff --git a/lxd/main_forksyscall.go b/lxd/main_forksyscall.go
index 8682406..669445d 100644
--- a/lxd/main_forksyscall.go
+++ b/lxd/main_forksyscall.go
@@ -220,8 +220,6 @@ static bool setnsat(int ns_fd, const char *ns)
 
 static bool change_creds(int ns_fd, cap_t caps, uid_t nsuid, gid_t nsgid, uid_t nsfsuid, gid_t nsfsgid)
 {
-	__do_close int fd = -EBADF;
-
 	if (prctl(PR_SET_KEEPCAPS, 1))
 		return false;
 
