diff --git a/install.sh b/install.sh
index 4f90d37..26e002f 100755
--- a/install.sh
+++ b/install.sh
@@ -3,6 +3,8 @@
 
 set -e
 
+: ${DESTDIR:=/usr/bin}
+
 if [ $# -eq 1 ] ; then
   if test -f bin/nim
   then
@@ -25,42 +27,16 @@ if [ $# -eq 1 ] ; then
       exit 1
       ;;
     "/usr/bin")
-      bindir=/usr/bin
-      configdir=/etc/nim
-      libdir=/usr/lib/nim
-      docdir=/usr/share/nim/doc
-      datadir=/usr/share/nim/data
-      nimbleDir="/opt/nimble/pkgs/compiler-1.2.0"
-      ;;
-    "/usr/local/bin")
-      bindir=/usr/local/bin
-      configdir=/etc/nim
-      libdir=/usr/local/lib/nim
-      docdir=/usr/local/share/nim/doc
-      datadir=/usr/local/share/nim/data
-      nimbleDir="/opt/nimble/pkgs/compiler-1.2.0"
-      ;;
-    "/opt")
-      bindir="/opt/nim/bin"
-      configdir="/opt/nim/config"
-      libdir="/opt/nim/lib"
-      docdir="/opt/nim/doc"
-      datadir="/opt/nim/data"
-      nimbleDir="/opt/nimble/pkgs/compiler-1.2.0"
-      mkdir -p /opt/nim
-      mkdir -p $bindir
-      mkdir -p $configdir
+      bindir=$DESTDIR/usr/bin
+      configdir=$DESTDIR/etc/nim
+      libdir="$DESTDIR/usr/lib/nim"
+      docdir="$DESTDIR/usr/share/doc/nim"
+      datadir="$DESTDIR/usr/share/nim/data"
+      nimbleDir="$DESTDIR/usr/share/nimble/pkgs"
       ;;
     *)
-      bindir="$1/nim/bin"
-      configdir="$1/nim/config"
-      libdir="$1/nim/lib"
-      docdir="$1/nim/doc"
-      datadir="$1/nim/data"
-      nimbleDir="$1/nim"
-      mkdir -p $1/nim
-      mkdir -p $bindir
-      mkdir -p $configdir
+      echo "nope"
+      exit 1
       ;;
   esac
 
