diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7d99ec3..c79fd41 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1024,7 +1024,7 @@ if(NOT DEFINED BUILD_SHARED_LIBS OR BUILD_SHARED_LIBS)
     set_target_properties(zlib PROPERTIES DEFINE_SYMBOL ZLIB_DLL)
 
     if(ZLIB_COMPAT)
-        set(ZLIB_FULL_VERSION ${ZLIB_HEADER_VERSION}.zlib-ng)
+        set(ZLIB_FULL_VERSION ${ZLIB_HEADER_VERSION})
         set_target_properties(zlib PROPERTIES SOVERSION 1)
     else()
         set(ZLIB_FULL_VERSION ${ZLIBNG_HEADER_VERSION})
