diff --git a/src/cmd/go/internal/work/init.go b/src/cmd/go/internal/work/init.go
index 473bd1a31b..7aedf2c7ba 100644
--- a/src/cmd/go/internal/work/init.go
+++ b/src/cmd/go/internal/work/init.go
@@ -106,6 +106,10 @@ func buildModeInit() {
 	// That way, if the flag is completely bogus we will prefer to error out with
 	// "-buildmode=%s not supported" instead of naming the specific platform.
 
+	if cfg.BuildBuildmode == "default" && cfg.Goos == "linux" && !cfg.BuildRace {
+		cfg.BuildBuildmode = "pie"
+	}
+
 	switch cfg.BuildBuildmode {
 	case "archive":
 		pkgsFilter = pkgsNotMain
