diff --git a/Makefile.am b/Makefile.am
index 40b300d..f2dbc73 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -4,7 +4,7 @@
 
 AUTOMAKE_OPTIONS = 1.9 gnu dist-bzip2 dist-xz check-news
 
-SUBDIRS = libpam tests libpamc libpam_misc modules po conf doc examples xtests
+SUBDIRS = libpam tests libpamc libpam_misc modules conf doc examples xtests
 
 CLEANFILES = *~
 
diff --git a/configure.ac b/configure.ac
index 3012ceb..4203f75 100644
--- a/configure.ac
+++ b/configure.ac
@@ -572,10 +572,6 @@ AM_CONDITIONAL(ENABLE_REGENERATE_MAN, test x$enable_docu != xno)
 AM_CONDITIONAL(ENABLE_GENERATE_PDF, test ! -z "$FO2PDF")
 
 
-AM_GNU_GETTEXT_VERSION([0.18.3])
-AM_GNU_GETTEXT([external])
-AC_CHECK_FUNCS(dngettext)
-
 AH_BOTTOM([#ifdef ENABLE_NLS
 #include <libintl.h>
 #define _(msgid) dgettext(PACKAGE, msgid)
@@ -609,7 +605,6 @@ AM_CONDITIONAL([HAVE_KEY_MANAGEMENT], [test "$have_key_syscalls" = 1])
 dnl Files to be created from when we run configure
 AC_CONFIG_FILES([Makefile libpam/Makefile libpamc/Makefile libpamc/test/Makefile \
 	libpam_misc/Makefile conf/Makefile conf/pam_conv1/Makefile \
-	po/Makefile.in \
 	modules/Makefile \
 	modules/pam_access/Makefile modules/pam_cracklib/Makefile \
         modules/pam_debug/Makefile modules/pam_deny/Makefile \
